action/protocol/rolldpos/epoch_test.go

```diff
@@ -1,36 +1,230 @@
+	receipt, error := p.Handle(ctx, nil, nil)
+	require.Nil(receipt)
+	require.NoError(error)
+}
+
+func TestProtocol_Validate(t *testing.T) {
+	require := require.New(t)
+	p := NewProtocol(23, 4, 3)
+	ctx := context.Background()
+	require.NoError(p.Validate(ctx, nil))
+}
+
+func TestProtocol_NumCandidateDelegates(t *testing.T) {
+	require := require.New(t)
+	p := NewProtocol(23, 4, 3)
+	require.Equal(uint64(23), p.NumCandidateDelegates())
+}
+
+func TestProtocol_NumDelegates(t *testing.T) {
+	require := require.New(t)
+	p := NewProtocol(23, 4, 3)
+	require.Equal(uint64(4), p.NumDelegates())
+}
+
+func TestProtocol_ReadState(t *testing.T) {
+	require := require.New(t)
+	p := NewProtocol(23, 4, 3)
+	ctx := context.Background()
+	methods := [8]string{
+		"NumCandidateDelegates",
+		"NumDelegates",
+		"NumSubEpochs",
+		"EpochNumber",
+		"EpochHeight",
+		"EpochLastHeight",
+		"SubEpochNumber",
+		"trick",
 	}
+
+	arg1 := byteutil.Uint64ToBytes(10)
+	arg2 := byteutil.Uint64ToBytes(20)
+
+	for i, method := range methods {
+
+		if i != 0 && i != 1 {
+			result, err := p.ReadState(ctx, nil, []byte(method), arg1, arg2)
+			require.Nil(result)
+			require.Error(err)
+		}
+
+		switch method {
+
+		case "NumCandidateDelegates":
+			result, err := p.ReadState(ctx, nil, []byte(method), arg1)
+			require.Equal(byteutil.Uint64ToBytes(p.numCandidateDelegates), result)
+			require.Nil(err)
+
+		case "NumDelegates":
+			result, err := p.ReadState(ctx, nil, []byte(method), arg1)
+			require.Equal(byteutil.Uint64ToBytes(p.numDelegates), result)
+			require.Nil(err)
+
+		case "NumSubEpochs":
+			result, err := p.ReadState(ctx, nil, []byte(method), arg1)
+			require.Equal(byteutil.Uint64ToBytes(p.NumSubEpochs(byteutil.BytesToUint64(arg1))), result)
+			require.Nil(err)
+
+		case "EpochNumber":
+
+			result, err := p.ReadState(ctx, nil, []byte(method), arg1)
+			require.Equal(byteutil.Uint64ToBytes(p.GetEpochNum(byteutil.BytesToUint64(arg1))), result)
+			require.Nil(err)
+
+		case "EpochHeight":
+
+			result, err := p.ReadState(ctx, nil, []byte(method), arg1)
+			require.Equal(byteutil.Uint64ToBytes(p.GetEpochHeight(byteutil.BytesToUint64(arg1))), result)
+			require.Nil(err)
+
+		case "EpochLastHeight":
+
+			result, err := p.ReadState(ctx, nil, []byte(method), arg1)
+			require.Equal(byteutil.Uint64ToBytes(p.GetEpochLastBlockHeight(byteutil.BytesToUint64(arg1))), result)
+			require.Nil(err)
+
+		case "SubEpochNumber":
+
+			result, err := p.ReadState(ctx, nil, []byte(method), arg1)
+			require.Equal(byteutil.Uint64ToBytes(p.GetSubEpochNum(byteutil.BytesToUint64(arg1))), result)
+			require.Nil(err)
+
+		default:
+			result, err := p.ReadState(ctx, nil, []byte(method), arg1)
+			require.Nil(result)
+			require.Error(err)
+
+		}
+
+	}
+
+}
+
+func TestProtocol_NumSubEpochs(t *testing.T) {
+
+	require := require.New(t)
+
+	height := []uint64{0, 1, 12, 25, 38, 53, 59, 80, 90, 93, 120}
+
+	expectedP := []uint64{3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3}
+
```

Comment:
TODO: set numSubEpochs = 3, numSubEpochsEnableDardanelles = 6. dardanellesHeight to 54.